### DISCUSSION TITLE
Change window title on Powershell?

### DISCUSSION THREAD

**Discussion by ransagy** [starship/starship#1259] [General]
As per the documentation, one needs to define the logic of setting the window title before launching Starship. However, The documentation only mentions the *nix family of shells.

In Powershell the way to do this AFAIK is the following:

```powershell
function SetWindowTitle {
    $host.UI.RawUI.WindowTitle = $PWD  # always set to the path, as an example.
}
```

However, I'm not too sure about the equivelant of the precmd commands in this case. Is this somethign Starship needs to do/support or am i missing something in my `$PROFILE` file to be set before Starship launches?

**Comment by jeanga** [comment]
I've been thinking about this and src/init/starship.ps1 probably needs to be modified to change the window title. No big a deal IMHO.
However, what value do we want the window title to be? the current working directory? the complete prompt?

**Reply by davidkna** [reply]
~~Shouldn't we just add documentation for that like with the nix shells?~~
Yes supporting some kind of `starship_pre_cmd ` thing sounds good.

**Comment by ransagy** [comment]
I'd start with the current working directory; I believe that's what Powershell does by default. Obviously other scripts/apps can change the title as well but that shouldn't concern us.

It would be a nice bonus to make it configurable as i can definitely imagine some people wanting it to mimic the prompt itself. (which in turn is exactly what would make it configurable to begin with?)

**Reply by davidkna** [reply]
I just tested in plain powershell and no window title gets set.

**Reply by ransagy** [reply]
Powershell 7 (Core) seems to do this for me; it could be the posh-git module i used before Starship, as well.

**Reply by davidkna** [reply]
Using `@(starship prompt)[1]` for the window title could work but you'd have to somehow get rid of the ansi color escapes (a second starship config with lots of style strings maybe?).

**Comment by davidkna** [comment]
> However, I'm not too sure about the equivelant of the precmd commands in this case. Is this somethign Starship needs to do/support or am i missing something in my `$PROFILE` file to be set before Starship launches?

In powershell you can use for now

```ps1
function SetWindowTitle {
    $host.UI.RawUI.WindowTitle = $PWD  # always set to the path, as an example.
}

# Starship must come first
starship init powershell | iex

$oldPrompt = $function:prompt
function prompt {
	$null = SetWindowTitle 
	& $oldPrompt
}
```
(stolen from [zoxide](https://github.com/ajeetdsouza/zoxide/blob/master/src/subcommand/init/shell/powershell.rs#L64-L70))

Also akin to zoxide for $PWD chnages only this will work:
```
$ExecutionContext.InvokeCommand.LocationChangedAction = {
	$null = SetWindowTitle 
}
```

**Reply by ransagy** [reply]
Seems to work for me without the last part. Maybe i misunderstood you though. Just adding the first part to my `$PROFILE` seems to work well.

**Reply by davidkna** [reply]
They are differnt methods to achive the same thing.
`$ExecutionContext.InvokeCommand.LocationChangedAction` triggers on every directory change and `prompt` on every prompt draw.

**Reply by ransagy** [reply]
Oh, I misread the last part, indeed. Either way, Thank you, works splendidly for now :]
